core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
+    /** Returns a description of the underlying source of the bytes, e.g. a path, stdin, a URL etc. */
+    String getDescription();
+
+    // Methods related to buffering/repositioning/skipping etc.
+    /** Returns true if the source has a known finite length, false otherwise. */
+    boolean hasLength();
```

Comment:
`hasLength` is a kind of funny name, maybe `hasKnownLength`?  More wordy though.

Reply:
Or `lenght` returning -1 if not known...

Reply:
I'd prefer not to use special values when possible.  I'm happy to change the name.  How would you feel about `Optional<Long> length()`?  `Option` is widely used in scala for this kind of thing, but I gather Java folks are not as keen on it?

Reply:
I vote for use `Optional` more in the htsjdk-next library as a way of represent returned data that might not be present. Much better than magic numbers or a getter method failing if not present and a boolean method to check (I hate to read again an again things similar to `hasLength ? length : 0`; optional in that case is more idiomatic).